src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -118,17 +127,22 @@ static ucs_status_t uct_cuda_copy_iface_flush(uct_iface_h tl_iface, unsigned fla
                                               uct_completion_t *comp)
 {
     uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_iface, uct_cuda_copy_iface_t);
+    ucs_queue_head_t *event_q;
 
     if (comp != NULL) {
         return UCS_ERR_UNSUPPORTED;
     }
 
-    if (ucs_queue_is_empty(&iface->outstanding_event_q[UCT_CUDA_COPY_STREAM_H2D]) &&
-        ucs_queue_is_empty(&iface->outstanding_event_q[UCT_CUDA_COPY_STREAM_D2H])) {
-        UCT_TL_IFACE_STAT_FLUSH(ucs_derived_of(tl_iface, uct_base_iface_t));
-        return UCS_OK;
-    }
+    uct_cuda_copy_for_each_event_q(iface, event_q, {
```

Comment:
this is not good - too much polling. we need to put all events into one queue to not poll too many locations.

Reply:
Currently there is a 1:1 mapping between event_q and stream per direction. Every time there is a cudamemcpy operation, we record an event and push it to the queue. Later in fd_arm, if one of the event_queues wasn't empty, we add a stream callback for the stream associated with that queue and let the callback notify ucp_worker_wait/epoll when the cudaMemcy call completes. If we unify all event_queues, then in fd_arm we may have to add stream callbacks to all streams because we don't know which stream is active. This would mean that potentially worker_wait will end up resolving false notifications and potentially causing an overhead as well.

Reply:
can we check which streams are active when arming the event? or keep a list of active stream?
in any case we should not poll on UCS_MEMORY_TYPE_LAST^2 locations during progress

Reply:
How about using another queue of active queues? We could check that queue instead of checking all queues. Will that work?

Reply:
yes it will be better